File: LeetCode/javascript/lc_0572.ts
// 暴力递归，每个节点都算一遍
function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {

    function compare(node: TreeNode | null, subNode: TreeNode | null) {
        if (node == null && subNode == null) {
            return true;
        }
        if (node && subNode) {
            if (node.val === subNode.val) {
                return compare(node.left, subNode.left) && compare(node.right, subNode.right);
            }
            return false;
        }
        return false;
    }

    let isSame = false;
    function traversal(node: TreeNode) {
        if (compare(node, subRoot)) {
            isSame = true;
        }
        else {
            if (node.left) traversal(node.left);
            if (node.right) traversal(node.right);
        }
    }

    root && traversal(root);

    return isSame;
};

File: LeetCode/cpp/lc_0572.cpp
/**
 * @file     lc_0572.cpp
 * @brief    https://leetcode-cn.com/problems/subtree-of-another-tree/
 * @author   YongDu
 * @date     2021-09-09
 */

class Solution {
public:
  bool isSubtree(TreeNode *root, TreeNode *subRoot) {
    if (subRoot == nullptr) {
      return true;
    }
    if (root == nullptr) {
      return false;
    }
    if (isSameTree(root, subRoot)) {
      return true;
    }
    return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
  }

private:
  bool isSameTree(const TreeNode *left, const TreeNode *right) {
    if (!left && !right) {
      return true;
    }
    if (!left || !right) {
      return false;
    }
    if (left->val != right->val) {
      return false;
    }
    return isSameTree(left->left, right->left) && isSameTree(left->right, right->right);
  }
};
